src/popup-page/modules/app.tsx

```diff
@@ -0,0 +1,47 @@
+import React from 'react'
+import { getItem, StoreObserver, setItem } from '../../utils/storage'
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface Props {}
+interface State {
+  beeApiUrl: string
+}
+
+class App extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+    this.state = { beeApiUrl: 'http://localhost:1633' }
+    this.asyncInit()
+
+    // bindings
+    this.handleBzzApiUrlChange = this.handleBzzApiUrlChange.bind(this)
+    this.handleSubmit = this.handleSubmit.bind(this)
+  }
+
+  handleBzzApiUrlChange(event: React.ChangeEvent<HTMLInputElement>): void {
+    this.setState({ beeApiUrl: event.target.value })
+  }
+
+  async handleSubmit(event: React.FormEvent<HTMLElement>): Promise<void> {
+    event.preventDefault()
+    await setItem('beeApiUrl', this.state.beeApiUrl)
+  }
+
+  async asyncInit(): Promise<void> {
+    this.setState({ beeApiUrl: await getItem('beeApiUrl') })
```

Comment:
While this works for simple cases on the long term I suggest using a state management library such as redux, redux-toolkit or mobx. Then the initialization of the state could happen when the application starts and then the required properties can be passed in as props.

Reply:
I don't think this will be such a huge library to require state management. I'd advocate to use the build in [Context](https://reactjs.org/docs/context.html) or if we want to be adventurous we could use [Recoil](https://github.com/facebookexperimental/Recoil) (facebook's state management with first class support for Rract )

Reply:
I made an issue about it with id #10